workers/unity/Packages/io.improbable.worker.sdk/IOStream.cs

```diff
@@ -26,95 +32,104 @@ public static IOStream CreateRingBufferStream(uint capacity)
             return new IOStream(CIO.CreateRingBufferStream(capacity));
         }
 
-        public static IOStream CreateFileStream(string fileName)
+        public static IOStream CreateFileStream(string fileName, OpenMode openMode)
         {
             fixed (byte* fileNameBytes = ApiInterop.ToUtf8Cstr(fileName))
             {
-                return new IOStream(CIO.CreateFileStream(fileNameBytes, CIO.OpenMode.OpenModeDefault));
+                return new IOStream(CIO.CreateFileStream(fileNameBytes, (CIO.OpenMode) openMode));
             }
         }
 
-        public void Write(byte[] data)
+        public long Write(byte[] data)
         {
             var remainingCapacity = CIO.StreamGetRemainingWriteCapacityBytes(stream);
             if (remainingCapacity < data.Length)
             {
                 throw new IOException("Not enough stream capacity to write data.");
             }
 
+            var bytesWritten = 0L;
             fixed (byte* dataToWrite = data)
             {
-                var bytesWritten = CIO.StreamWrite(stream, dataToWrite, 1);
-
-                if (bytesWritten == -1)
-                {
-                    // The Stream write failed, check the error message
-                    var rawError = CIO.StreamGetLastError(stream);
-                    throw new IOException(ApiInterop.FromUtf8Cstr(rawError));
-                }
-
-                if (bytesWritten != data.Length)
-                {
-                    // Check that the number of bytes we tried to write was the number of bytes written
-                    throw new IOException("Failed to write all data to stream.");
-                }
+                bytesWritten = CIO.StreamWrite(stream, dataToWrite, 1);
             }
+
+            if (bytesWritten != -1)
+            {
+                return bytesWritten;
+            }
+
+            var rawError = CIO.StreamGetLastError(stream);
+            throw new IOException(ApiInterop.FromUtf8Cstr(rawError));
         }
 
-        public byte[] Read(uint bytesToRead)
+        public (byte[], long) Read(uint bytesToRead)
```

Comment:
Not sure if we want to use a tuple, or an `out` var here.
@jamiebrynes7 ?

Reply:
Probably an `out` parameter like: `public long Read(uint bytesToRead, out bytes[] bytes)` or similar